extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/TransactionManagerBuildTimeConfig.java

```diff
@@ -0,0 +1,34 @@
+package io.quarkus.narayana.jta.runtime;
+
+import io.quarkus.runtime.annotations.ConfigItem;
+import io.quarkus.runtime.annotations.ConfigPhase;
+import io.quarkus.runtime.annotations.ConfigRoot;
+
+@ConfigRoot(phase = ConfigPhase.BUILD_TIME)
+public final class TransactionManagerBuildTimeConfig {
+    /**
+     * Allow using multiple XA unaware resources in the same transactional demarcation with XA capable resources.
+     * This is UNSAFE and may only be used for compatibility.
+     * <p>
+     * Either use XA for all resources if you want consistency, or split the code into separate
+     * methods with separate transactions.
+     * <p>
+     * Note that using a single XA unaware resource together with XA aware resources, known as
+     * the Last Resource Commit Optimization (LRCO), is different from using multiple XA unaware
+     * resources. Although LRCO allows most transactions to complete normally, some errors can
+     * cause an inconsistent transaction outcome. Using multiple XA unaware resources is not
+     * recommended since the probability of inconsistent outcomes is significantly higher and
+     * much harder to recover from than LRCO. For this reason, use LRCO as a last resort.
+     * <p>
+     * We plan to remove this property in the future so you should plan fixing your application
+     * accordingly.
+     * If you think your use case of this feature is valid and this option should be kept around,
```

Comment:
Hmm, I'll need to keep a look out for anyone that takes advantage of this invitation!

Reply:
Don't worry, I'll ping all of you if/when it occurs so that we can discuss it.

Reply:
Keycloak is hitting this situation because it automatically wraps all requests in a transaction, defaults to using a non-xa primary datasource, and allows users to contribute extensions that are typically not documented as requiring xa - we've had users using the Amazon Web Services JDBC Driver which specifically lacks support for XA. 

@mmusgrov In nearly all our logic paths multiple resources will not be updated together. If allowUnsafeMultipleLastResources will we get something like a huerstic mixed exception that we can clearly reference in our docs. And/or is there some why in which we or extension developers could indicate a datasource is read-only (Connection.setReadOnly? - although that will fail if used against an XA datasource Connection right) for the purposes of this check?

cc @pedroigor @vmuzikar

Reply:
> Keycloak is hitting this situation because it automatically wraps all requests in a transaction, defaults to using a non-xa primary datasource, and allows users to contribute extensions that are typically not documented as requiring xa - we've had users using the Amazon Web Services JDBC Driver which specifically lacks support for XA.

@shawkins do you have a way to detect these cases and wrap them in a separate transaction with `QuarkusTransaction.requiringNew().run(() -> {})`/`QuarkusTransaction.requiringNew().call(() -> {})`? Or have the extensions do this themselves? This would get rid of the error/warning, and as far as I understand would give you similar behavior to what you currently get when you don't use XA. @mmusgrov should be able to confirm.

Of course, you'd still be in a transactionally unsafe situation, since you're not using XA. But if you can't use XA, that's about the best you can do, I think.

> @mmusgrov In nearly all our logic paths multiple resources will not be updated together. If allowUnsafeMultipleLastResources will we get something like a huerstic mixed exception that we can clearly reference in our docs

Not sure what you mean by "huerstic mixed exception", but hopefully  @mmusgrov will.

> And/or is there some why in which we or extension developers could indicate a datasource is read-only (Connection.setReadOnly? - although that will fail if used against an XA datasource Connection right) for the purposes of this check?

From what I read in other messages, even if all your datasources but one are read-only, it is transactionally unsafe not to use XA. Don't ask me why, maybe @mmusgrov can explain.

Reply:
> If allowUnsafeMultipleLastResources will we get something like a huerstic mixed exception that we can clearly reference in our docs.

I don't know since Narayana does not test multiple last resources.

Reply:
Since it is only recommended for developer mode/prototyping and being transactionally unsafe we never envisaged it being used in production code and therefore we don't have unit tests. I'm really not sure what we would put in such a unit test. This property is a workaround for applications that didn't realise they were exposed prior to the Agroal change.

Reply:
> do you have a way to detect these cases and wrap them in a separate transaction with QuarkusTransaction.requiringNew().run(() -> {...})/QuarkusTransaction.requiringNew().call(() -> {...})? Or have the extensions do this themselves?

Must of the update paths that I'm seeing target a single source - that is if a federated user is stored in source x, we're only updating x. Read paths, such as searching for users, can look across sources. There may be situations there where multiple calls are made to the same source - it would be nice if we didn't have to refactor to ensure those were under the same transaction. 

There are a couple of exceptions to this - granting a role to all users, and there are hooks where an extension developer could do anything. It would be far easier to demarcate the paths that possibly could update multiple sources.

> From what I read in other messages, even if all your datasources but one are read-only, it is transactionally unsafe not to use XA. Don't ask me why, maybe @mmusgrov can explain.

There's an explanation here: https://jbossts.blogspot.com/2018/06/narayana-commit-markable-resource.html

Reply:
There is LRCO which, although still unsafe, allows one XA unaware resource to participate in an XA transaction. With LRCO the XA unaware resource is processed last in the prepare phase of the transaction, at which time an attempt is made to commit it. If the system crashes before we have written a transaction log record (which we do if all XA aware resources prepared and if the XA unaware resource committed) then the system looses consistency because, on restart, when recovery runs we won't have the log so we use "presumed abort" and we tell each XA resource to rollback (ie one resource will have committed whilst others will have rolled back).

Reply:
> This property is a workaround for applications that didn't realise they were exposed prior to the Agroal change ...

Note that prior to the Agroal change the user was not given any indication that they were using multiple last resources so the situation was even more problematic (at least now the problem is exposed giving users a chance to plan for the failure case).

Reply:
Okay, we've had similar discussions in the past, e.g. https://github.com/quarkusio/quarkus/issues/6414#issuecomment-1148818020 .

I'll add a comment on that other issue, maybe let's continue the discussion there @shawkins ? This doesn't look like something that could be solved on the spot anyway, but rather a new feature.

Reply:
@yrodiere sounds good.

Reply:
Alright @shawkins  so your problems are not just about read-only. Let's continue the discussion here then.

[Your message](https://github.com/quarkusio/quarkus/issues/6414#issuecomment-2091439188):

> @yrodiere thank you for the proposal. The current Keycloak logic directly uses the jta TransactionManager to control transaction boundaries - but we can of course adapt as needed.
> 
> A base level ask mentioned on the pr was for a heurstic mixed exception that would help users understand if something unsafe has happened during a commit.
> 
> Beyond that maybe some of our concern could be addressed by tweaking the meaning of the transactions property. That is if you specify - quarkus.datasource.name.jdbc.transactions=enabled - then it should behave like agroal 2.1 meaning best effort 1pc. That would provide backwards compatibility.
> 
> If you specify - quarkus.datasource.name.jdbc.transactions=xa - and only a non-XA driver is available or specified, then rather than there being a configuration error it should behave like agroal 2.3. That is create a local xa resrouce, but allow only 1 of them per transaction. If this is too subtle of a change, another property value could be introduced, maybe something like "safe"?
> 
> The docs could then explain why "enabled" is just best effort and potentially unsafe with multiple datasources.
> 
> On top of that some consideration of readonly would be great to have - it would either refine either the commit exception handling for the enabled case (we'd know whether it was a heurstic mixed situation), or otherwise allow for multiple non-xa resources even though safe handling is expected.

Reply:
> A base level ask mentioned on the pr was for a heurstic mixed exception that would help users understand if something unsafe has happened during a commit.

You do get an exception in that case, it's just really confusing. I opened #40431, but TBH this is probably not something we want to fix in Quarkus itself. We need a fix in Agroal.

> Beyond that maybe some of our concern could be addressed by tweaking the meaning of the transactions property. That is if you specify - quarkus.datasource.name.jdbc.transactions=enabled - then it should behave like agroal 2.1 meaning best effort 1pc. That would provide backwards compatibility.

We already intend to provide backwards compatibility for 3.8 -- where we'll backport this PR *and* set `quarkus.transaction-manager.allow-unsafe-multiple-last-resources` to `true` by default, leading to something like Agroal 2.1, with a few warnings for unsafe situations.

But if that solution is not good enough, please explain how.

> If you specify - quarkus.datasource.name.jdbc.transactions=xa - and only a non-XA driver is available or specified, then rather than there being a configuration error it should behave like agroal 2.3. That is create a local xa resrouce, but allow only 1 of them per transaction. If this is too subtle of a change, another property value could be introduced, maybe something like "safe"?

I don't think hijacking the existing configuration values is a good idea, but it could make sense to add a new value (`xa-if-available`?) here that would detect whether to use `xa` or not based on the driver. Or a more global flag that would tell Quarkus "when transactions are `enabled` (not `xa`), prefer XA if possible, but use whatever the driver allows otherwise". However:

1. Let's clarify: you want this feature in order to be able to nudge all datasources (even those from Keycloak extensions) towards XA by default?
2. Have you tried setting an XA driver explicitly and not setting `quarkus.datasource.name.jdbc.transactions`, or setting it to `enabled`? I suspect it already works exactly like you mentioned.
  I'll grant your that defaulting e.g. all postgresql datasources to XA is not something you can do in Quarkus right now, however. I'm not sure it's a good idea though, as that requires explicit configuration on the database side.
3. This is probably not something we'll solve in this PR, so if you think this PR is not enough,we'll probably need a separate issue to track what the problem is and discuss how to solve it.

Reply:
> But if that solution is not good enough, please explain how.

The thinking is that you don't have to introduce something like quarkus.transaction-manager.allow-unsafe-multiple-last-resources=true if you just make that behavior part of what "enabled" means, but if you prefer a global flag that's ok too.

> Let's clarify: you want this feature in order to be able to nudge all datasources (even those from Keycloak extensions) towards XA by default?

No, it's to localize the transactional expectation of the 1pc resource rather than to rely on a global flag.

In our case we don't yet have xa support baked fully into our container based deployments - we haven't yet added stable storage for the transaction log, so even if users enable xa they are not getting the full benefit unless they do additional work.

> Have you tried setting an XA driver explicitly and not setting quarkus.datasource.name.jdbc.transactions, or setting it to enabled? I suspect it already works exactly like you mentioned.

If you set the driver explicitly to an XA datasource and leave the transactions as enabled, then yes I believe it will still behave as a full XA resource. The only difference is that recovery would not be used for that source. So currently "enabled" means resolve to the non-XA datasource a driver is not specified and don't use recovery.

"xa" currently means resolve to the XA datasource if one is not specified and use recoverty.

I was looking for an in-between. Expect "safe" XA semantics, but use the non-XA driver if that's all that is available - which is what Agroal 2.3 seems to be enforcing by default.

Taking a step back, the situation with keycloak is that XA is disabled by default - and we claim support for at least 1 non-XA database. If the user does not install any extensions with their own datasource(s), that works quite well. If the user installs an extension that users additional resources, then we're hitting the case that initiated an issue - that used to "work" whether or not you configured the main keycloak datasource, or the additional resources as xa.  That was expected to be safe in all but 1 code paths that I've found (there's one that adds a role to all users - across federated sources). And even if the user wanted to use xa recovery wasn't available in a container - unless they did additional work as mentioned above. So I'm not sure Keycloak really considered true XA support a critical feature.

Reply:
> The thinking is that you don't have to introduce something like quarkus.transaction-manager.allow-unsafe-multiple-last-resources=true if you just make that behavior part of what "enabled" means, but if you prefer a global flag that's ok too.

I think the global flag will have to do for now, because that whole patch is about reverting a regression in 3.8 by patching 3.8.next. In that context:

1. Semantics changes are a no-go.
2. We need to merge the patch soon-ish.
3. ... so there is little time to properly discuss a new feature and its impact.
4. ... and there is little time to work with upstream (Narayana) to change their SPIs -- which would likely be required if we wanted to do more than provide a *global* flag.

So, if that's fine with you (from what you wrote, it seems to be), we'll go with this solution -- which, for the record, is expected to be temporary -- and will work on a dedicated feature in another issue. Maybe you could open it, adding an example of how you see it used in Keycloak?

Reply:
Sounds good, I'll open another issue and reference it here.